src/Microsoft.AspNet.OData.Shared/Formatter/Serialization/ODataResourceSerializer.cs

```diff
@@ -634,6 +638,109 @@ public virtual ODataResource CreateResource(SelectExpandNode selectExpandNode, R
             }
         }
 
+        /// <summary>
+        /// Method to append InstanceAnnotations to the ODataResource and Property
+        /// </summary>
+        /// <param name="resource">param for ODataResource</param>
+        /// <param name="resourceContext">param for ODataResourceContext</param>        
+        public virtual void AppendInstanceAnnotations(ODataResource resource, ResourceContext resourceContext)
+        {
+            IEdmStructuredType structuredType = resourceContext.StructuredType;
+            IEdmStructuredObject structuredObject = resourceContext.EdmObject;
+            PropertyInfo instanceAnnotationInfo = EdmLibHelpers.GetInstanceAnnotationsDictionary(structuredType,
+                resourceContext.EdmModel);
+
+            object value;
+            
+            if (instanceAnnotationInfo == null || structuredObject == null ||
+                !structuredObject.TryGetPropertyValue(instanceAnnotationInfo.Name, out value) || value == null)
+            {
+                return;
+            }                  
+
+            IDictionary<string, IDictionary<string, object> > instanceAnnotationsDictionary = value as IDictionary<string, IDictionary<string, object>>;
+
+            if(instanceAnnotationsDictionary != null)
+            {
+                if (instanceAnnotationsDictionary.ContainsKey(string.Empty))
+                {
+                    if (resource.InstanceAnnotations == null)
+                    {
+                        resource.InstanceAnnotations = new List<ODataInstanceAnnotation>();
+                    }
+
+                    foreach (KeyValuePair<string,object> annotation in instanceAnnotationsDictionary[string.Empty])
+                    {
+                        IEdmTypeReference edmTypeReference = resourceContext.SerializerContext.GetEdmType(annotation.Value,
+                           annotation.Value.GetType());
+                         
```

Comment:
Probably should handle the case where `annotation.Value` is null. Should we still serialize the annotation in that case? Maybe - we could set the value to `new ODataNullValue()`

Reply:
Yes, we should support setting an annotation value to null. 

@Sreejith - can you add a test case for setting an annotation value to null?

---
In reply to: [473388044](https://github.com/OData/WebApi/pull/2219#discussion_r473388044) [](ancestors = 473388044)